atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraAbsentHostTracker.java

```diff
@@ -16,22 +16,66 @@
 
 package com.palantir.atlasdb.keyvalue.cassandra;
 
+import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
 import com.palantir.atlasdb.keyvalue.cassandra.pool.CassandraServer;
+import com.palantir.common.concurrent.PTExecutors;
 import com.palantir.common.streams.KeyedStream;
 import com.palantir.logsafe.SafeArg;
 import com.palantir.logsafe.logger.SafeLogger;
 import com.palantir.logsafe.logger.SafeLoggerFactory;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
 import javax.annotation.concurrent.GuardedBy;
+import one.util.streamex.EntryStream;
 import org.immutables.value.Value;
 
 public final class CassandraAbsentHostTracker {
     private static final SafeLogger log = SafeLoggerFactory.get(CassandraAbsentHostTracker.class);
+    private static final AtomicLong ELEMENTS_MADE_ABSENT = new AtomicLong(0);
+    private static final AtomicLong REMOVED_ELEMENTS = new AtomicLong(0);
+    private static final Set<CassandraAbsentHostTracker> KNOWN_TRACKERS = ConcurrentHashMap.newKeySet();
+    private static final ScheduledExecutorService SCHEDULED_EXECUTOR_SERVICE =
+            PTExecutors.newSingleThreadScheduledExecutor();
+
+    static {
+        SCHEDULED_EXECUTOR_SERVICE.scheduleWithFixedDelay(
```

Comment:
do we need to track the returned future and cancel it on close/shutdown?

Reply:
We definitely should if this was meant to go into prod code (otherwise when we close a TM we get a thread leak), though given this was never meant to be merged I was a bit lazy